New readme template

tforster opened the pull request:
Updated template to reflect discussion from earlier today with @webvidelement.

tforster committed:
Updated README template and externalised CHANGELOG

tforster committed:
Bumped version numbers

webdivelement commented on README.md:
```diff
@@ -4,79 +4,136 @@
 
 Troy Forster – [@tforster](https://twitter.com/tforster) – troy.forster@gmail.com
 
 Distributed under the ISC license. See `LICENSE` for more information.
```
Does this LICENSE file exist yet?

tforster replied:
It does not. It's probably a good habit to get into, adding an actual license file along with the simple mention of a license in package.json. It is another "control" file in the root though. I'm ok with that because it serves a unique purpose. It's not quite the same has having a katrillion dot-files all vying for some linting config attention.

What say you to adding LICENSE to our template. We could have two, one is our typical permissive OSS license (actual choice tbd) and the other we would reserve for private client projects. LICENSE-OSS.md and LICENSE-CLOSED.md and the instructions say delete one, rename the other.

webdivelement replied:
That sounds good to me.

webdivelement commented on README.md:
````diff
@@ -4,79 +4,136 @@
+```
 
-- 0.1.0 **Initial Creation** (2020-04-09)
+## Change log
````
This should be title case, ex: `Change Log`.
```suggestion
## Change Log
```

tforster replied:
Sorry, you're becoming my proof-reader. Yes, that should be `Change Log`.

webdivelement replied:
No problem, that's what a second set of eyes is for.

webdivelement commented on README.md:
```diff
@@ -4,79 +4,136 @@
 Inspired by Eric Meyers CSS Reset, Git New is a simple tool to enforce a consistent baseline Git experience for new projects.
 
-## Installation
+- [Git New](#git-new)
```
Is the top level heading necessary?  Is it auto-generated?

tforster replied:
Are you referring to the ToC top level? Yes, auto-generated. It's the challenge of applying web style headings to a document style MD file. Another decision we need to make. In a document, you can have multiple heading 1's. Each one corresponding to a chapter (or similar). Mostly because a non-web document has pages, whereas a web page is, well, a page. Some of my MDs can be quite long, especially architecture, etc. I would prefer to use multiple H1s in an MD, treating it like a document that just happens to be web-renderable.

Thoughts?

webdivelement replied:
I am cool with multiple heading 1's on the page. It makes semantic sense, the same way you can have multiple H1s in HTML5 as long as they are in appropriate regions.

webdivelement commented on README.md:
```diff
@@ -4,79 +4,136 @@
+- [Git New](#git-new)
 
-Linux, Windows and OSX
+  - [For Project Developers](#for-project-developers)
```
The headings / anchors here seem to be inconsistent with the content below.

tforster replied:
Hmm, should be auto-update on save. I will confirm.

webdivelement commented on README.md:
```diff
@@ -4,79 +4,136 @@
+Follow this section if you will be developing new features and fixing bugs for this repository.
 
-Development setup is very easy. Just clone this repository and start editing.
+### Prerequisites
```
See notes above in #prerequisites section.

webdivelement commented on README.md:
```diff
@@ -4,79 +4,136 @@
+
+### Prerequisites
+
+- A good code editor. We suggest [VSCode](https://code.visualstudio.com/), [Atom](https://atom.io/) or [SublimeText](https://www.sublimetext.com/).
```
Although I like promoting good tools, is the IDE necessary for the README? The project can be run, edited, maintained, built, deployed, etc. all without the IDEs listed. Despite the incredible usefulness of the tools, I wouldn't say they are necessarily required the way Node or Git is. Thoughts?

tforster replied:
I dunno why I keep coming back to this. Yes, let's pull out the IDEs and editors.
```suggestion
- A good code editor.
```

webdivelement commented on README.md:
```diff
@@ -4,79 +4,136 @@
+### Prerequisites
+
+- A good code editor. We suggest [VSCode](https://code.visualstudio.com/), [Atom](https://atom.io/) or [SublimeText](https://www.sublimetext.com/).
+- [NodeJS LTS+ and NPM 6.14+](https://nodejs.org/en/download/)
```
Should we specify versions for things like Node to future proof it? I don't know if patch or even minor is necessary, but I would think major might be handy, as things could change down the road. Thoughts?

tforster replied:
What's your view on the LTS convention? When I work (assuming I am not constrained by AWS Lambda runtimes) I tend to go for the latest Node version. And I will live with the consequences of being on the breaking edge. But for collab/shared source should we stick with latest and just list the version? Or favour stability and recommend LTS?

webdivelement replied:
I like favouring the current LTS for shared source or client projects.  However, simply listing LTS doesn't help when we try to run the project a few years later as there could be a different LTS version.  In the future, if I knew it was a specific major (even minor) version, I could use NVM to switch before running it.  This mostly applies to projects that come and go quickly and don't see long term development, where the code base would evolve as the tools do.

tforster replied:
However, if we pin to a version (use Node as an example) and someone clones the repo 2 years later does that imply they have to downgrade their version to one from 2 years prior? 

I was adding a plus sign (e.g. NPM 6.14+) but that gets a bit redundant. 

```suggestion
### Prerequisites

The versions listed for these prerequisites are current at the time of writing. More recent versions will likely work but "your mileage may vary".

- A good code editor. 
- [Node.js v12.13.0 and NPM 6.14.5](https://nodejs.org/en/download/)
```

tforster replied:
FYI, I'm using ctrl-g in GitHub to insert suggested changes that can be accepted and merged from this page (it's pretty cool). But in the above one I notice it is not replacing the full text, only the line :(

webdivelement replied:
I think it depends on what changes over time, which is out of our control.  Pinning to a major or even minor version may guarantee that it will run on that version, where future major versions may intro breaking changes that is incompatible with our code.  Most of our projects will be fine running in the future, but there is a chance that it doesn't run on Node 20.  I think adding that line that you did makes the most sense.

webdivelement replied:
I see what you mean, it duplicated the content in the commit.

webdivelement commented on CHANGELOG.md:
```diff
@@ -0,0 +1,11 @@
+- Also externalised this change log to CHANGELOG.md
+
+v0.1.1 **Updated eslint dependency version** (2020-05-29)
+
```
Should we always include at least one descriptive line below the version/title or should we omit it if it will be identical to the title?

tforster replied:
I struggled with that too and decided that omitting is best. Some changes, e.g. majors and minors should get a decent description. A patch is typically very small and can likely be summarised in just the title/sentence (referencing my earlier comment questioning whether we need titles or just go with a descriptive sentence).

webdivelement replied:
Sounds good, omitting it is.

webdivelement reviewed (commented):
I added some notes to discuss. I wasn't sure if I should make edits inline or discuss first.

tforster committed:
Update README.md

tforster committed:
Update README.md

Case typo

Co-authored-by: webdivelement <webdivelement@users.noreply.github.com>

tforster committed:
Update README.md

tforster committed:
Added a new permisive license file that implements MIT
Reference MIT license in package.json
Improve formatting and accuracy of README.md

tforster commented:
@webdivelement I have added a LICENSE.txt file that implements the permissive MIT license. Instructions are now included in the README to use the LICENSE, or choose a different one and update the license property in package.json. I also moved the "For Developers" section to "Contributing" at the bottom of the file and added a new "For Users" placeholder section for any repos that need to describe how to download, install, etc (thinking this is where instructions to npm i the repo would go).

We're getting there. Take a look and comment.

tforster commented:
After re-reading it in place on GitHub with GitHub formatting I think we can collapse For Developers and For Users into simply "Usage"? Thoughts?

webdivelement commented:
I think that makes sense as well, to keep it concise. Let's see how that looks.

tforster committed:
Collapsed For Developers into Usage

tforster commented:
@webdivelement So that meant we had usage/usage which was easily fixed by moving everything up to heading 1 level, which also simplifies the reading.

I think we're there. My only annoyance that I am not fixing right now is that the amazing tiny VSCode extension that gives me MD snippets, including auto-TOC, won't recognise the doc title and TOC heading, which I feel should not be part of the TOC.  TOC referenced inside the TOC is very PHP-meta feeling. 

Hoping you feel this is final thumbs-up worthy too.
